### PR TITLE
[GNA] Sinking Transpose through Split and Concat without changing axis

### PR DESCRIPTION
**Pull request by evkotov**
### Details:
 - add TSSplitBackward and TSConcatForward transformations

### Tickets:
 - 93857
- 80034
- 82144
- 93873
- 102239
- 105056
- 105090
- 106327

### CODE REVIEW

File: src/plugins/intel_gna/src/transformations/utils/transformation_helper.cpp [hunk 1]
```diff
@@ -153,7 +153,8 @@ void remove_single_input_node(std::shared_ptr<ov::Node> node) {
     if (!node_parent) {
         THROW_GNA_EXCEPTION << "The removing node has no parrent node";
     }
-    if (!std::equal(input_node_shape.begin(), input_node_shape.end(), output_node_shape.begin())) {
+    if (input_node_shape.size() != output_node_shape.size() ||
```
Comment (sirzabek) [inline src/plugins/intel_gna/src/transformations/utils/transformation_helper.cpp:156]: With this change I think we should rename the function to remove_single_input_output_node.

File: src/plugins/intel_gna/tests/unit/transformations/gather_sinking_test_utils.cpp [hunk 2]
```diff
@@ -0,0 +1,50 @@
+// Copyright (C) 2023 Intel Corporation
```
Comment (sirzabek) [inline src/plugins/intel_gna/tests/unit/transformations/gather_sinking_test_utils.cpp:1]: Functions in this file have already been added in gather sinking PR, please watch out not to duplicate and also follow the comments in the other PR.

File: src/plugins/intel_gna/tests/unit/transformations/gather_sinking_test_utils.hpp [hunk 3]
```diff
@@ -0,0 +1,31 @@
+// Copyright (C) 2023 Intel Corporation
```
Comment (sirzabek) [inline src/plugins/intel_gna/tests/unit/transformations/gather_sinking_test_utils.hpp:1]: Functions in this file have already been added in gather sinking PR, please watch out not to duplicate and also follow the comments in the other PR.

**Review by sirzabek — APPROVED** [review]
As agreed approved with comments.

File: src/plugins/intel_gna/src/common/graph_utils.hpp [hunk 4]
```diff
@@ -374,6 +374,18 @@ inline std::vector<size_t> make_gather_indices_from_transpose_axes(const Shape&
     return gather_order;
 }
 
+/**
+ * @brief Finds first dimension != 1
+ */
+inline size_t get_first_valuable_dim_id(const ov::Shape& shape) {
```
[the remark this answers is not in the document]
Comment (evkotov) [reply src/plugins/intel_gna/src/common/graph_utils.hpp:380]: This function is needed in current PR and cannot been built without it. Another PR is not merged into master.

File: src/plugins/intel_gna/src/transformations/ts_concat.cpp [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+using namespace ov::opset12;
+using namespace ov::pass::pattern;
+using namespace ov::intel_gna;
+using namespace ov::intel_gna::pass;
+using namespace ov::intel_gna::pass::helper;
+using namespace ov::intel_gna::limitations;
+
+namespace {
+
+bool is_concat_sinked(const Output<Node>& output) {
```
[the remark this answers is not in the document]
Comment (evkotov) [reply src/plugins/intel_gna/src/transformations/ts_concat.cpp:27]: Sure. We discussed it offline. It should be done in a separate PR

File: src/plugins/intel_gna/tests/unit/transformations/gather_sinking_test_utils.hpp (hunk 3 above)
Comment (evkotov) [reply src/plugins/intel_gna/tests/unit/transformations/gather_sinking_test_utils.hpp:1]: This functions are needed in current PR and cannot been built without it. Another PR is not merged into master.

File: src/plugins/intel_gna/tests/unit/transformations/gather_sinking_test_utils.cpp (hunk 2 above)
Comment (evkotov) [reply src/plugins/intel_gna/tests/unit/transformations/gather_sinking_test_utils.cpp:1]: This functions are needed in current PR and cannot been built without it. Another PR is not merged into master.

File: src/plugins/intel_gna/src/transformations/utils/transformation_helper.cpp (hunk 1 above)
Comment (evkotov) [reply src/plugins/intel_gna/src/transformations/utils/transformation_helper.cpp:156]: I don't see any huge changes. We just befor std::equal additionally check if compared vectors has different sizes. It is needed for Windows build - Windows STL throws unneeded exception in a such case

File: src/plugins/intel_gna/src/transformations/ts_split.cpp [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+            auto transpose = ov::as_type_ptr<Transpose>(input.get_node()->shared_from_this());
+            if (transpose && !Limitations::is_transpose_supported(transpose))
+                return true;
+        }
+    }
+    return false;
+}
+
+std::vector<size_t> create_gather_indices(const ov::Shape& input_shape, const ov::Shape& order) {
+    if (input_shape.size() < 2 || input_shape.size() > 4) {
```
[the remark this answers is not in the document]
Comment (evkotov) [reply src/plugins/intel_gna/src/transformations/ts_split.cpp:40]: In a general situation we should do all the checks inside patcher pass pattern. But here we have a complex situation, that cannot be proceeded by ngraph pattern: split can have different count of child nodes and any of them can be Transpose. So we can find a Split node and than search and check child Transposes (that is done here) or Transpose and search for it parent Split

File: src/plugins/intel_gna/src/transformations/ts_split.cpp [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+    // common case when shape is 4d
+    std::vector<size_t> xyz_4d = {input_shape_4d[3] * input_shape_4d[2] * input_shape_4d[1],
+                                  input_shape_4d[3] * input_shape_4d[2],
+                                  input_shape_4d[3],
+                                  1};
+
+    std::vector<size_t> xyz = transpose_shape(xyz_4d, order_4d);
+    std::vector<size_t> gather_order;
+
+    for (size_t n = 0; n < output_shape_4d[0]; ++n) {
```
Comment (mryzhov) [inline src/plugins/intel_gna/src/transformations/ts_split.cpp:62]: How many times are we going to duplicate this code? please use make_gather_indices_from_transpose_axes

File: src/plugins/intel_gna/src/transformations/ts_split.cpp [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+        auto gather_indices =
+            std::make_shared<Constant>(ov::element::i64, ov::Shape{gather_indices_value.size()}, gather_indices_value);
+        auto gather = std::make_shared<Gather>(reshape_input, gather_indices, gather_axis);
+
+        auto split_axis_new = std::make_shared<Constant>(ov::element::i64, ov::Shape{}, 1);
+        auto split_new = std::make_shared<Split>(gather, split_axis_new, split_node->get_num_splits());
+
+        ov::copy_runtime_info(split_node, {gather_axis, gather_indices, gather, split_axis_new, split_new});
+
+        for (size_t output_idx = 0; output_idx < split_node->get_output_size(); ++output_idx) {
```
Comment (mryzhov) [inline src/plugins/intel_gna/src/transformations/ts_split.cpp:147]: Please use the same code as for Concat

File: src/plugins/intel_gna/tests/unit/transformations/gather_sinking_test_utils.cpp [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    std::copy(vec.end() - k, vec.end(), buffer.begin());
+
+    for (int i = vec.size() - 1 - k; i >= 0; --i) {
+        vec[i + k] = vec[i];
+    }
+
+    std::copy(buffer.begin(), buffer.end(), vec.begin());
+}
+
+std::vector<size_t> gather_forward(size_t size, size_t initial_value) {
```
Comment (mryzhov) [inline src/plugins/intel_gna/tests/unit/transformations/gather_sinking_test_utils.cpp:38]: do we use this function in the PR?

File: src/plugins/intel_gna/src/transformations/ts_concat.cpp [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+    auto concat_node = ov::as_type_ptr<Concat>(output.get_node_shared_ptr());
+
+    const Shape concat_output_shape = concat_node->get_output_shape(0);
+    const int64_t axis = concat_node->get_concatenation_axis();
+    if (graph_utils::get_first_valuable_dim_id(concat_output_shape) != axis)
```
[the remark this answers is not in the document]
Comment (mryzhov) [reply src/plugins/intel_gna/src/transformations/ts_concat.cpp:32]: Could you provide an examples?

File: src/plugins/intel_gna/src/transformations/ts_split.cpp [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+            for (auto& input : split_node->get_output_target_inputs(output_idx)) {
+                auto transpose = ov::as_type_ptr<Transpose>(input.get_node()->shared_from_this());
+                if (transpose && !Limitations::is_transpose_supported(transpose)) {
```
Comment (mryzhov) [inline src/plugins/intel_gna/src/transformations/ts_split.cpp:150]: transpose can be supported on one of the outputs, this is incorrect check

File: src/plugins/intel_gna/tests/unit/transformations/gather_sinking_test_utils.cpp (hunk 9 above)
Comment (evkotov) [reply src/plugins/intel_gna/tests/unit/transformations/gather_sinking_test_utils.cpp:38]: Ok. I removed it

File: src/plugins/intel_gna/src/transformations/utils/transformation_helper.cpp (hunk 1 above)
Comment (sirzabek) [reply src/plugins/intel_gna/src/transformations/utils/transformation_helper.cpp]: I didn't say there are huge changes, but that with these changes it makes sense to rename the function, because it's what it actually does. Anyway now it's really hard to follow this and other PRs, because they are interdependent and the code is quickly outdated.

File: src/plugins/intel_gna/src/transformations/ts_concat.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (mryzhov) [reply src/plugins/intel_gna/src/transformations/ts_concat.cpp:27]: just for future, imho this is sinking specific function and it would be better do not move it to common graph utils

File: src/plugins/intel_gna/src/transformations/ts_concat.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (mryzhov) [reply src/plugins/intel_gna/src/transformations/ts_concat.cpp:32]: That is a different level of the checking, for this transformation it is important to check that axis is not supported by default. I know only one case when it can be supported - when Concat inputs connected directly to network inputs, but this is not a pattern of this transformation.

**Review by mryzhov — APPROVED** [review]
